bin/cylc-check-software

```diff
@@ -186,10 +186,11 @@ def cmd_find_ver(module, min_ver, cmd_base, ver_opt, ver_extr, outfile=1,
             res = [NOTFOUND_MSG, False]
         else:
             try:
-                output = procopen([cmd, ver_opt], stdoutpipe=True,
-                                  stdin=open(os.devnull),
-                                  stderrpipe=True).communicate()
-                [outfile - 1].decode().strip()
+                output = procopen(
+                    [cmd, ver_opt], stdoutpipe=True,
+                    stdin=open(os.devnull),
+                    stderrpipe=True).communicate()[outfile - 1].decode()\
+                    .strip()
```

Comment:
Not a problem of this change.

The `outfile` setting here is a bit awkward. The normal interface looks like `out, err = proc.communicate()`, so I guess we are looking for `0` or `1` for index here? So we are expecting `outfile` to be `1` (for STDOUT) or `2` (for STDERR)?

Just not very clear unless you know `subprocess.Popen` very well.

Reply:
I agree that's a bit hard to understand. `git blame` says most recently touched by Martin's bandit changes, but the construct dates back to before that, so perhaps @sadielbartholomew can comment?

Reply:
(But I guess the actual bug here was introduced by @MartinRyan (and missed by our reviews - sorry).

Reply:
@matthewrmshin 

>Just not very clear unless you know subprocess.Popen very well.

+1, I re-read the code a few times to try to understand what it was doing. Then had to open the docstring of `communicate`

> (But I guess the actual bug here was introduced by @MartinRyan (and missed by our reviews - sorry).

Yup. But we will miss more small issues like this. But as long as we can cover code like this with simple tests, we should be good :grimacing: 

Should I create a separate, low priority ticket, to review the way we are handling Popen, communicate, and accessing the array response in this method? Or would it be best to address that here now? I'd prefer to first fix this syntax issue, and later see how to simplify this code.

Reply:
I'm pretty sure this is a one-off unusual use of Popen (plus @MartinRyan has recently wrapped Popen to assuage bandit, so there are few distinct calls now - all go via the wrapper). So yeah, just fix this syntax issue I'd say.

Reply:
Sorry, I missed this namecheck until now.

I'll be honest, I do not have much more understanding of the motivator & context behind the ``outfile`` setting, as I based a lot of my code for ``check-software`` on that written by @oliver-sanders for [``rose check-software``](https://github.com/metomi/rose/blob/93a23f98c8521cafbfcc4377da9325ac9aea557b/bin/rose-check-software#L109-L110), & though I will have understood it superficially, I recall I struggled to comprehend what the numerical ``outfile``-based index was doing too (but left it in because it worked :grin: ). Perhaps Oliver can comment further?